independent-projects/parent/pom.xml

```diff
@@ -696,6 +691,267 @@
+                                            <property>
+                                                <name>removeUnused</name>
+                                            </property>
+                                            <property>
+                                                <name>treatSamePackageAsUnused</name>
+                                            </property>
+                                            <property>
+                                                <name>breadthFirstComparator</name>
+                                            </property>
+                                            <property>
+                                                <name>lineEnding</name>
+                                            </property>
+                                            <property>
+                                                <name>compliance</name>
+                                            </property>
+                                        </properties>
+                                        <ignoredProperties>
+                                            <ignore>project</ignore>
+                                            <ignore>plugin</ignore>
+                                        </ignoredProperties>
+                                    </inputs>
+                                    <outputs>
+                                        <!-- For now we don't want to output the cachedir as it contains absolute paths
+                                        See https://github.com/revelc/impsort-maven-plugin/pull/87
+                                        <directories>
+                                            <directory>
+                                                <name>cachedir</name>
+                                            </directory>
+                                        </directories>
+                                        -->
+                                        <cacheableBecause>check should only run when inputs change
+                                        </cacheableBecause>
+                                    </outputs>
+                                </plugin>
+                                <plugin>
+                                    <groupId>org.apache.maven.plugins</groupId>
+                                    <artifactId>maven-enforcer-plugin</artifactId>
+                                    <inputs>
+                                        <fileSets>
+                                            <fileSet>
+                                                <name>foo</name>
+                                                <paths>
+                                                    <path>${project.basedir}/pom.xml</path>
+                                                </paths>
+                                                <normalization>RELATIVE_PATH</normalization>
+                                            </fileSet>
```

Comment:
@jprinet I used your proposal except I restricted it to the pom file as really that's what all our enforcer rules are checking.

That being said, I have two questions:
- are the plugin dependencies taken into account in the cache key? We have rules that are inside a jar declared as a dependency. If the jar changes, will the cache be invalidated?
- some of our rules are inspecting the transitive dependencies so I would need to somehow include the classpath into the cache key. Is there a way to do that when the plugin doesn't expose it?

I saw the following for the kotlin-maven-plugin but these are properties from the plugin:

```
                                            <fileSet>
                                                <name>classpath</name>
                                                <normalization>COMPILE_CLASSPATH</normalization>
                                            </fileSet>
                                            <fileSet>
                                                <name>testClasspath</name>
                                                <normalization>COMPILE_CLASSPATH</normalization>
                                            </fileSet>
```

Is there a way to include the classpath jars into the cache key?

Reply:
> are the plugin dependencies taken into account in the cache key? We have rules that are inside a jar declared as a dependency. If the jar changes, will the cache be invalidated?

There is no default input, the plugin has to define it which is not the case apparently for [the enforcer plugin](https://github.com/apache/maven-enforcer/blob/enforcer-3.4.1/maven-enforcer-plugin/src/main/java/org/apache/maven/plugins/enforcer/EnforceMojo.java)

> some of our rules are inspecting the transitive dependencies so I would need to somehow include the classpath into the cache key. Is there a way to do that when the plugin doesn't expose it?

The only way would be with the programmatic approach, nothing you can do with the XML configuration afaik.
We are doing [that](https://github.com/gradle/develocity-build-config-samples/blob/a0027c4807b4bdb4b9f9d3085f2a4c3304ec6618/quarkus-build-caching-extension/src/main/java/com/gradle/QuarkusBuildCache.java#L278) for the Quarkus build caching extension.